stats/src/main/java/io/airlift/stats/JmxGcMonitor.java

```diff
@@ -0,0 +1,267 @@
+{
+    private final Logger log = Logger.get(JmxGcMonitor.class);
+
+    private final NotificationListener notificationListener = (notification, ignored) -> onNotification(notification);
+
+    private final AtomicLong majorGcCount = new AtomicLong();
+    private final AtomicLong majorGcTime = new AtomicLong();
+    private final TimeStat majorGc = new TimeStat();
+
+    private final TimeStat minorGc = new TimeStat();
+
+    @GuardedBy("this")
+    private long lastGcEndTime = System.currentTimeMillis();
+
+    @PostConstruct
+    public void start()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().addNotificationListener(
+                        objectName,
+                        notificationListener,
+                        null,
+                        null);
+            }
+            catch (JMException e) {
+                throw new RuntimeException("Unable to add GC listener", e);
+            }
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().removeNotificationListener(objectName, notificationListener);
+            }
+            catch (JMException ignored) {
+            }
+        }
+    }
+
+    @Override
+    public long getMajorGcCount()
+    {
+        return majorGcCount.get();
+    }
+
+    @Override
+    public Duration getMajorGcTime()
+    {
+        return new Duration(majorGcTime.get(), MILLISECONDS);
+    }
+
+    @Managed
+    @Nested
+    public TimeStat getMajorGc()
+    {
+        return majorGc;
+    }
+
+    @Managed
+    @Nested
+    public TimeStat getMinorGc()
+    {
+        return minorGc;
+    }
+
+    private synchronized void onNotification(Notification notification)
+    {
+        if ("com.sun.management.gc.notification".equals(notification.getType())) {
+            GarbageCollectionNotificationInfo info = new GarbageCollectionNotificationInfo((CompositeData) notification.getUserData());
+
+            if (info.isMajorGc()) {
+                majorGcCount.incrementAndGet();
+                majorGcTime.addAndGet(info.getDurationMs());
+                majorGc.add(info.getDurationMs(), MILLISECONDS);
+
+                // assume that major GCs stop the application
+                long applicationRuntime = max(0, info.getStartTime() - lastGcEndTime);
+                lastGcEndTime = info.getEndTime();
+
+                log.info(
+                        "Major GC: application %sms, stopped %sms:: %s -> %s",
+                        applicationRuntime,
+                        info.getDurationMs(),
+                        info.getBeforeGcTotalMb(),
```

Comment:
We should either use DataSize for these or add the appropriate unit in the log message.